### DISCUSSION TITLE
Exemple with FR and my problem

### DISCUSSION THREAD

**Discussion by iAmTosuj** [csells/go_router#304] [Q&A] [answered]
Hello, I made an example, for these issue https://github.com/csells/go_router/issues/279 and  https://github.com/csells/go_router/issues/282.

my example https://github.com/iAmTosuj/go_router_exemple, I left comments in the code.

I described everything in the readme and will duplicate it here.

### Feature request.
 The application has authentication by code and deep link. I open the application using the link and I need to get to it after I enter the authentication code. To do this, I need to remember the link before authentication. It would be convenient if I received not only a string, but also the name of this link, so it would be easier to check it. You can see an example in the router_redirect file.

in the state Router, it would be possible to transmit the entire route history in a flat list. In my case, I would take the last matched route and check its name and the name of my deeplinks. Then you don't need to write regular expressions for location params

### Strange behavior.
The page is recreated every time you return to it or use hot restart. But the strangest thing is that all final fields are created anew, and the old fields remain in widget. You can see an example with bloc in the main_page file.

Before that, I used the auto_router library and there was no such problem, the constructor was never called when going back through history or when use hot restart.

i can do this, but how can I change the parameters of this page if they are passed to the constructor?

```dart
final mainPage = MainPage();

GoRoute(path: '/main', builder: (context, state) => mainPage,)
````
You can make it so that if the page has already been created, it will be cached and when you return to it, a new instance will not be created, it will be taken from the cache?

Maybe I'm doing something wrong and I'm having these problems. Please help if possible.

**Comment by csells** [comment, marked as answer]
You can ask Flutter to cache the state of any widget you like -- that's what a `StatefulWidget` is. Put the state that you want kept between invocations into your State<T> object and override `didUpdateWidget` for new data in the associated `widget`. You can read more here: https://flutterbyexample.com/lesson/stateful-widget-lifecycle

**Reply by iAmTosuj** [reply]
thanks for the reply. Today I found this solution, but it is not convenient because there are many such pages and you will need to do this every time. But at the moment I don't see any other options.

**Comment by iAmTosuj** [comment]
what about adding route history to the root redirect for GoRouter, is it possible? or is there such a need only for me?

**Comment by csells** [comment]
What kind of route history do you mean? Are you looking for a list of everywhere the user has been in the app? You can get that today by passing in a `NavigatorObserver` implementation. In fact, I believe that @esdotdev has built such a thing and may be willing to share.

**Reply by esDotDev** [reply]
I actually don't use `NavigatorObserver` to do this, as there are cases where the path can change, but `NavigatorObserver` does not fire, as it relies soley on the page-stack to trigger changes. I logged a bug on this once iirc, but not sure if it ever got solved.

GR is a change notifier, so in the past I just `addListener` to it, and record locations to build up a history. I find it more straight-forward and robust compared to using NavigatorObserver.

**Reply by iAmTosuj** [reply]
from my example

```dart
 redirect: (state) {
      final AuthStateState _authState = Get.find<AuthStateCubit>().state;
      final SavedLinkCubit _savedLinkCubit = Get.find<SavedLinkCubit>();

      // save deeplink for open current page after auth
      // I save only nested routes, it would be convenient to get the state of the root route here
      // if you remove the link check, there will be an error
      if (state.location.startsWith('/main/')) {
        _savedLinkCubit.saveLink(state.location);
      }

      if (!_authState.isAppInit) {
        return state.location != '/splash' ? '/splash' : null;
      }

      if (!_authState.isUserAuth) {
        return state.location != '/auth' ? '/auth' : null;
      }

      // clear saved link if user logged
      _savedLinkCubit.clearLink();

      return state.location == '/' ? '/main' : null;
    },
```
I use redirect and I need to save the deeplink if I didn't follow it right away
    
The problem is that I need to save only some links, otherwise there will be an error and if instead of the condition `state.location.startsWith('/main/')`, I wrote `state.name == route.name`, then it would be easier to extend the code.

```dart
    routes: [
      GoRoute(
        path: '/splash',
        builder: (context, state) => SplashPage(),
      ),
      GoRoute(
        path: '/auth',
        builder: (context, state) => const AuthPage(),
      ),
      GoRoute(path: '/main', builder: (context, state) => MainPage(), routes: [
        GoRoute(
          path: 'child',
          builder: (context, state) => const ChildPage(),
        ),
      ]),
    ],
 ```
    
 if I want to switch to `/main/child`. Then in the redirect method for GoRouter I would like to get a list of routes in the state. `[MainRouteState, ChildRouteState]`. And then I would take the last route, get its name and compare it with the expected route name.
 
Perhaps you can make an additional field where the state of all matched routes will be located

**Comment by csells** [comment]
You can get a list of matching routes today from `router.routerDelegate.matches`.
